### PR TITLE
Allow usage of [<TailCall>] with older FSharp.Core package versions

### PR DESCRIPTION
**Pull request by dawedawe**
## Description

This allows users to use the `[<TailCall>]` analysis with older versions of `FSharp.Core`. They just need to define the attribute in `Microsoft.FSharp.Core` themselves.

## Checklist

- [X] Test cases added
- [ ] Performance benchmarks added in case of performance changes
- [x] Release notes entry updated

### CODE REVIEW

File: src/Compiler/Checking/TailCallChecks.fs [hunk 1]
```diff
@@ -56,6 +56,10 @@ type TailCall =
         | TailCall.Yes _ -> TailCall.Yes TailCallReturnType.NonVoid
         | TailCall.No -> TailCall.No
 
+let hasTailCallAttrib (attribs: Attribs) =
+    attribs
+    |> List.exists (fun a -> a.TyconRef.CompiledRepresentationForNamedType.FullName = "Microsoft.FSharp.Core.TailCallAttribute")
+
```
Comment (vzarytovskii) [inline src/Compiler/Checking/TailCallChecks.fs:62]: This will probably need to be in TcGlobals as other attributes and have usual checks as well (tryFindAttribute and co). 

Like we do with IsReadOnlyAttribute (for required attributes as well iirc) for example. You can even do exactly like that, and compiler will generate TailCallAttribute if it doesn't find it in ccus.

**Comment by vzarytovskii** [comment]
Just a note - while it's a valid scenario, we highly discourage people from using custom FSharp.Core.

**Review by T-Gro — APPROVED** [review]

File: src/Compiler/TypedTree/TcGlobals.fs [hunk 2]
```diff
@@ -1852,6 +1851,10 @@ type TcGlobals(
 
   member _.DebuggerNonUserCodeAttribute = debuggerNonUserCodeAttribute
 
+  member _.HasTailCallAttrib (attribs: Attribs) =
+    attribs
+    |> List.exists (fun a -> a.TyconRef.CompiledRepresentationForNamedType.FullName = "Microsoft.FSharp.Core.TailCallAttribute")
+
```
Comment (vzarytovskii) [inline src/Compiler/TypedTree/TcGlobals.fs:1857]: Can't we use something like findPublicSysAttrib or tryFindSysAttrib? Like we do with RequiredAttribute for example. One of them _should_ be able to find them in all ccus

**Comment by smoothdeveloper** [comment]
What I like here is, like in C# which has those attributes, they can be added as internal in older projects that don't reference the recent SDK, and the compiler just look at those the same as if they were baked in the SDK. (for C#, I'd prefer record and nullable support to be shipped to .net framework consumers though...).

I don't think it is about "custom version of FSharp.Core" but just playing nice with consumers that can't easily (or just haven't done the) upgrade to latest one, if this comes at a very minor cost.

Taking the opposite approach to extreme (not saying this is what is done, F# is overall better than wider .net on this front, IME), even if it is "2 years were given", I see it a bit the same as "python 2 to 3", but this is so long it is very minor cost in the compiler impl.

If something is possible for string interpolation with FSharp.Core < 5, it may be worth a look, and to keep in mind for other attributes used by the compiler.

File: src/Compiler/TypedTree/TcGlobals.fs (hunk 2 above)
Comment (dawedawe) [reply src/Compiler/TypedTree/TcGlobals.fs:1857]: I tried the following:
```fsharp
  member _.HasTailCallAttrib (attribs: Attribs) =
      let tailCallAttr = findSysAttrib "Microsoft.FSharp.Core.TailCallAttribute"
      List.exists
              (fun (Attrib(tyconRef = tcref2)) ->
                  tcref2.CompiledRepresentationForNamedType.FullName = tailCallAttr.TyconRef.CompiledRepresentationForNamedType.FullName)
              attribs
```
But that raises the following:
```
FSC : error FS0074: The type referenced through 'Microsoft.FSharp.Core.TailCallAttribute' is defined in an assembly that is not referenced. You must add a reference to assembly 'The system type 'Microsoft.FSharp.Core.TailCallAttribute' was required but no referenced system DLL contained this type'. [C:\Users\schae\src\tailcallaltat 
tr\tailcallaltattr.fsproj]
```
As `tryFindSysAttrib` also uses `tryFindSysTypeCcuHelper` internally, that doesn't work either.

To better illustrate the usecase, here's a [repo](https://github.com/dawedawe/tailcallaltattr), that uses FSharp.Core 6.0.7 and gets TailCall warnings with this PR.

**Comment by vzarytovskii** [comment]
> What I like here is, like in C# which has those attributes, they can be added as internal in older projects that don't reference the recent SDK, and the compiler just look at those the same as if they were baked in the SDK. (for C#, I'd prefer record and nullable support to be shipped to .net framework consumers though...).
> 
> 
> 
We do the same for implicit attributes. Can't do with explicit ones.


> I don't think it is about "custom version of FSharp.Core" but just playing nice with consumers that can't easily (or just haven't done the) upgrade to latest one, if this comes at a very minor cost.
> 
> 
> 
The only situation where it's needed is when newer compiler is using older fslib. And the only way of doing is to upgrade sdk but specifically ping core lib to old one. And that's generally discouraged.

**Comment by smoothdeveloper** [comment]
> The only situation where it's needed is when newer compiler is using older fslib. And the only way of doing is to upgrade sdk but specifically pin core lib to old one. And that's generally discouraged.

I may misunderstand, but I think the encouraged thing is:
* library pin to lower FSharp.Core, per what they need (not forcing consumers to update due to a dependency)
* applications / leaf projects that are close to make runtime decision have the latitude to pick any FSharp.Core version, so long the compiler supports it

We encourage to support old for "inner / dependencies", or new for "close to runtime".

In this context, one library may decide to remain on lower FSharp.Core but define couple of internal attributes for their own consumption of features of recent F# compiler; it would generally not affect consumers of the library that may still use FSharp.Core that doesn't or does have those attributes.

File: src/Compiler/TypedTree/TcGlobals.fs (hunk 2 above)
Comment (vzarytovskii) [reply src/Compiler/TypedTree/TcGlobals.fs:1857]: It bothers me slightly that we just do string comparison in the list of attributes, instead of checking as we usually do. 

It's not the hill I'm willing to die on though, if everyone else is fine with it.

Comment (dawedawe) [reply src/Compiler/TypedTree/TcGlobals.fs:1875]: Thanks for the reviews so far.
But, uhm, what's the state here: Can we go on with this or are changes required?
If the string comparison is not okay, can you explain what could go wrong?

Comment (vzarytovskii) [reply src/Compiler/TypedTree/TcGlobals.fs:1875]: Two things - we don't really do it usually, and by checking the name we don't actually check it's a proper attribute (as attribute functions ensure).

That said, I'm fine with merging it, I can fix it later.

File: tests/FSharp.Compiler.ComponentTests/ErrorMessages/TailCallAttribute.fs [hunk 3]
```diff
@@ -1486,3 +1486,30 @@ namespace N
               Message =
                            "The TailCall attribute should only be applied to recursive functions." }
         ]
+
+    [<FSharp.Test.FactForNETCOREAPP>]
+    let ``Warn about self-defined attribute`` () = // is the analysis available for uers of older FSharp.Core versions
```
Comment (psfinaki) [inline tests/FSharp.Compiler.ComponentTests/ErrorMessages/TailCallAttribute.fs:1491]:
```suggestion
    let ``Warn about self-defined attribute`` () = // is the analysis available for users of older FSharp.Core versions
```

**Review by psfinaki — APPROVED** [review]
Hmm I am also not a fan of the string comparison but since I don't have a better idea and this seems to be quite an isolated piece of functionality (and also edge-cased for the older F# core), I am fine with letting this in.

File: tests/FSharp.Compiler.ComponentTests/ErrorMessages/TailCallAttribute.fs (hunk 3 above)
Comment (psfinaki) [inline tests/FSharp.Compiler.ComponentTests/ErrorMessages/TailCallAttribute.fs]: Also, stupid question (I am not very strong at F# component tests) - how does this test verify the older version of F# core? It still requires "withLangVersionPreview".

Comment (dawedawe) [reply tests/FSharp.Compiler.ComponentTests/ErrorMessages/TailCallAttribute.fs]: You are right, this test doesn't use an older `FSharp.Core` package (e.g. v6.0.7). As far as I know, we don't have infrastructure for such tests and the `withLangVersionPreview` doesn't affect the used `FSharp.Core` version.

The reason why I still think that this test is valid is the precedence that the locally defined Attribute takes in this test.
For example, if you rename the locally defined Attribute to `Foo` and adjust the comparison function in `TcGlobals.fs` accordingly, the test still passes.
To make this a bit more obvious, I removed the "withLangVersionPreview" from the test.

Comment (vzarytovskii) [reply tests/FSharp.Compiler.ComponentTests/ErrorMessages/TailCallAttribute.fs]: Can you please make a separate example project inside `tests/projects` which uses locally built compiler and older FSharp.Core, and check that it works as expected?

Comment (dawedawe) [reply tests/FSharp.Compiler.ComponentTests/ErrorMessages/TailCallAttribute.fs]: Sure, I added one and checked that it works as expected.

Comment (vzarytovskii) [reply tests/FSharp.Compiler.ComponentTests/ErrorMessages/TailCallAttribute.fs]: Thanks, I'm slightly surprised that adding only `<DotnetFscCompilerPath>` was enough. It wasn't before.

**Review by vzarytovskii — APPROVED** [review]
